### DISCUSSION TITLE
pyodide install

### DISCUSSION THREAD

**Discussion by raybellwaves** [anitagraser/movingpandas#238] [General]
I'm motivated on making this package installable via pyodide. Going to use this thread to capture notes along the way

Default way to install things for pyodide is micropip (https://pyodide.org/en/stable/usage/api/micropip-api.html). Given movingpandas isn't pip installable see below there may be other options.

Use jupyterlite readthedocs website for testing:
https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html

Open a new notebook with the pyodide kernal and try

```
import micropip
await micropip.install("movingpandas")
```

This gives

```
ValueError: Can't find a pure Python 3 wheel for 'rasterio<=1.0.13'.
See: https://pyodide.org/en/stable/usage/faq.html#micropip-can-t-find-a-pure-python-wheel
You can use `micropip.install(..., keep_going=True)`to get a list of all packages with missing wheels.
```
https://pyodide.org/en/stable/usage/faq.html#micropip-can-t-find-a-pure-python-wheel

```
await micropip.install("movingpandas", keep_going=True)
```

gives
```
ValueError: Can't find a pure Python 3 wheel for: 'rasterio<=1.0.13', 'fiona>=1.8', 'rasterio'
See: https://pyodide.org/en/stable/usage/faq.html#micropip-can-t-find-a-pure-python-wheel
```

There is also piplite https://jupyterlite.readthedocs.io/en/latest/quickstart/using.html#a-python-kernel-powered-by-pyodide that comes with pyolite (https://blog.jupyter.org/jupyterlite-jupyter-%EF%B8%8F-webassembly-%EF%B8%8F-python-f6e2e41ab3fa) which is a wrapper to micropip. not sure on the difference

```
import piplite
await piplite.install("movingpandas")
```

which gives the same message

**Comment by raybellwaves** [comment]
I'm aware via https://github.com/anitagraser/movingpandas/blob/master/README.md#installation

installing movingpandas isn't recommended via pip due to the cartopy dependency.

(Note that it is NOT recommended to install MovingPandas from [PyPI](https://pypi.org/project/movingpandas/)! If you're on Windows or Mac, many GeoPandas / MovingPandas dependencies cannot be pip installed (for details see the corresponding notes in the [GeoPandas documentation](https://geopandas.readthedocs.io/en/latest/getting_started/install.html)). On Ubuntu, pip install fails on cartopy with "Proj 4.9.0 must be installed".)

```
Collecting cartopy
Downloading ... Cartopy-0.20.3
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [2 lines of output]
      <string>:117: UserWarning: Unable to determine GEOS version. Ensure you have 3.7.2 or later installed, or installation may fail.
      Proj 8.0.0 must be installed.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Reply by raybellwaves** [reply]
See https://github.com/anitagraser/movingpandas/issues/78 for minimizing deps in movingpandas e.g. plotting ones

**Reply by raybellwaves** [reply]
See https://github.com/SciTools/cartopy/issues?q=is%3Aissue+is%3Aopen+pip for issues with installing cartopy using pip

**Comment by raybellwaves** [comment]
I remember @dopplershift doing some work on getting https://github.com/Unidata/MetPy installable in the pyodide ecosystem. I need to dig it up

**Reply by dopplershift** [reply]
Since cartopy is optional, for MetPy we only had to deal with PyProj (which is now included in pyodide) and pooch (pyodide now can install _lzma to fix import of pooch).

Since Shapely and geos are in pyodide now, Cartopy *should* be do-able, but I haven't tried.

**Reply by raybellwaves** [reply]
@doppleshift 's PyProj PR to pyodide: https://github.com/pyodide/pyodide/pull/2555

**Comment by raybellwaves** [comment]
Looking at pip requirements https://github.com/anitagraser/movingpandas/blob/master/setup.py#L7 here is what is currently on pyodide

numpy: Y
matplotlib: Y
shapely: Y
pandas: Y
geopandas: N (Can't find a pure Python 3 wheel for: 'fiona>=1.8')
hvplot: Y
bokeh: Y
cartopy: N (Can't find a pure Python 3 wheel for: 'cartopy')
geoviews: N (Can't find a pure Python 3 wheel for: 'cartopy>=0.18.0')
pyproj: Y
geopy: Y

**Reply by raybellwaves** [reply]
See https://github.com/pyodide/pyodide/issues/1569 for geopandas and further up https://github.com/pyodide/pyodide/issues/3091 for fiona

**Comment by raybellwaves** [comment]
https://pyodide.org/en/stable/console.html

```
import micropip

# probably don't need all of these 
await micropip.install(["numpy", "matplotlib", "shapely", "pandas", "hvplot", "bokeh", "pyproj", "geopy"])
await micropip.install("geopandas", deps=False)
await micropip.install("movingpandas", deps=False)

import movingpandas as mpd
```
gives
```
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python3.10/site-packages/movingpandas/__init__.py", line 5, in <module>
    from .trajectory import *
  File "/lib/python3.10/site-packages/movingpandas/trajectory.py", line 6, in <module>
    import contextily as ctx
ModuleNotFoundError: No module named 'contextily'
```

```
import pandas as pd
df = pd.read_json(micropip.freeze())
df.loc["movingpandas"].loc["packages"]{'name': 'movingpandas', 'version': '0.1.dev2', 'file_name': 'https://files.pythonhosted.org/packages/32/f0/5a022cab35b5c03d137284a15ef11f929db13057f4b806cf36a4
d93e360c/movingpandas-0.1.dev2-py3-none-any.whl', 'install_dir': 'site', 'sha256': 'd7b2b41e0cec254560be29ca51dcff3f76dd89f9f8514e76af62a5af7d298c69', 'imports'
: ['movingpandas', 'tests'], 'depends': []}
```

**Reply by raybellwaves** [reply]
micorpip brings in 0.1.dev2...

I belive this is the last version with a wheel (https://pypi.org/project/movingpandas/0.1.dev1/#files)

**Reply by raybellwaves** [reply]
```
pip install numpy matplotlib shapely pandas hvplot bokeh pyproj geopy
pip install geopandas --no-deps
pip install movingpandas --no-deps
python
import movingpandas as mpd
mpd.__version__
```
works

**Comment by MarcSkovMadsen** [comment]
You can now install movingpandas and dependencies with micropip. See https://github.com/pyodide/pyodide/issues/3248

Unfortunately there is another issue

![image](https://github.com/user-attachments/assets/430753d9-a0ec-4048-bd87-90c07261559e)

FYI. @anitagraser @raybellwaves
